### PR TITLE
Fix accuracy issues with Gemma models

### PR DESCRIPTION
**Pull request by kunal-vaishnavi**
### Description

This PR fixes accuracy issues with Google's Gemma models by using bfloat16 precision, [always using float32 precision to compute any LayerNorms](https://github.com/huggingface/transformers/blob/fee1190601b5d04ec6d3f7f58fd22788d7f3236d/src/transformers/models/gemma3/modeling_gemma3.py#L141-L146), and casting the output logits to float32 always.

### Motivation and Context

This PR has been tested with Gemma-2 and Gemma-3. It is using the bfloat16 changes from [this PR](https://github.com/microsoft/onnxruntime-genai/pull/1447) as well as the missing final norm changes from [this PR](https://github.com/microsoft/onnxruntime-genai/pull/1420).

### CODE REVIEW

File: src/python/py/models/builder.py [hunk 1]
```diff
@@ -1065,6 +1106,74 @@ def make_layernorm(self, layer_id, layernorm, skip, simple, location):
...
+        self.layernorm_attrs["output_0"] = output_0
+
+        if raw_out3 is not None:
+            self.make_node("Cast", inputs=[raw_out3], outputs=[output_3], name=f"{name}/CastOutput3", to=self.io_dtype,)
+            self.make_value_info(output_3, self.io_dtype, shape=["batch_size", "sequence_length", self.hidden_size],)
+            self.layernorm_attrs["output_3"] = output_3
+            self.layernorm_attrs["root_input"] = output_3
+
+    def make_layernorm(self, layer_id, layernorm, skip, simple, location):
+        if "gemma-3" in self.model_name_or_path and self.io_dtype == TensorProto.BFLOAT16:
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:1172]: Let's have a setting added to [`layernorm_attrs`](https://github.com/microsoft/onnxruntime-genai/blob/89a288984f860b7a6cb0436d3989630a55a8ab0f/src/python/py/models/builder.py#L160) for casting to/from float32 precision instead of having two separate functions. The setting can be set to false by default and set to true for the Gemma models in the Gemma class constructors (for example, how `add_offset` is set).

File: src/python/py/models/builder.py [hunk 2]
```diff
@@ -1263,12 +1379,23 @@ def make_qk_norm(self, layer_id, attention):
...
         self.make_reshape(q_reshape_1_name, q_reshape_1_inputs, dtype=self.io_dtype, shape=['batch_size', 'sequence_length * num_attention_heads', self.head_size])
 
-        # Make Q LayerNorm
         q_layernorm_name = f"/model/layers.{layer_id}/attn/q_norm/SimplifiedLayerNormalization"
         q_weight_name = f"model.layers.{layer_id}.attn.q_norm.layernorm.weight"
         q_layernorm_output = f"{q_layernorm_name}/output_0"
-        self.make_external_tensor(attention.q_norm.weight.detach().numpy().astype(self.to_numpy_dtype[self.io_dtype]) + self.layernorm_attrs["add_offset"], q_weight_name)
-        self.make_node("SimplifiedLayerNormalization", inputs=[q_reshape_1_output, q_weight_name], outputs=[q_layernorm_output], name=q_layernorm_name, **layernorm_kwargs)
+
+        if "gemma-3" in self.model_name_or_path and self.io_dtype == TensorProto.BFLOAT16:
```
[the remark this answers is not in the document]
Comment (nenad1002) [reply src/python/py/models/builder.py:1386]: There is no easy separation of logic for this because make_layernorm creates new names and value infos, pulling all of that to calling methods would cause a mess.

The logic can be merged by having methods make_layernorm_fp32_helper and make_layernorm_helper that other methods can call.

[the remark this answers is not in the document]
Comment (nenad1002) [reply src/python/py/models/builder.py:1386]: There is no easy way to separate logic as original make_layernorm actually creates naming and value infos. I'll create some helper methods to make this easier, and make_qk_norm can call directly into those methods.

File: src/python/py/models/builder.py [hunk 3]
```diff
@@ -100,9 +100,12 @@ def __init__(self, config, io_dtype, onnx_dtype, ep, cache_dir, extra_options):
 
         # Map output names to their types and shapes
         self.output_names = ["logits"]
+
+        logits_type = TensorProto.FLOAT if "gemma-3" in self.model_name_or_path else self.io_dtype
+
         self.output_types = {
             "hidden_states": self.io_dtype,                                                                      # For standard models where you want to remove the language modeling head from the model (note that `hidden_states` is written this way to match Hugging Face format)
-            "logits": self.io_dtype,                                                                             # For standard models
+            "logits": logits_type,                                                                               # For standard models
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:108]: Let's create a method called `make_outputs_init()` for this. The `make_outputs_init` method can then be defined in the Gemma class and we can override the logits dtype there. Here's an example.

https://github.com/microsoft/onnxruntime-genai/blob/e24337c8cdd01d08415dd78fd026958ee269af2e/src/python/py/models/builder.py#L2750-L2753

File: src/python/py/models/builder.py [hunk 4]
```diff
@@ -816,8 +819,16 @@ def make_matmul_float(self, matmul, name, root_input, **kwargs):
...
         output = "logits" if kwargs.get("logits", False) else f"{name}/output_0"
-        self.make_node("MatMul", inputs=[root_input, weight], outputs=[output], name=name)
-        self.make_value_info(output, self.io_dtype, shape=['batch_size', 'sequence_length', last_dim])
+
+        if output != "logits" or "gemma-3" not in self.model_name_or_path:
+            self.make_node("MatMul", inputs=[root_input, weight], outputs=[output], name=name)
+            self.make_value_info(output, self.io_dtype, shape=['batch_size', 'sequence_length', last_dim])
+        else:
+            cast_name = name + "/Cast"
+            cast_name_input = name + "/Cast"
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:828]: Let's have another `make_*` function for MatMul + Cast instead.

File: src/python/py/models/builder.py [hunk 5]
```diff
@@ -1308,9 +1397,16 @@ def make_qk_norm(self, layer_id, attention):
         k_layernorm_name = f"/model/layers.{layer_id}/attn/k_norm/SimplifiedLayerNormalization"
         k_weight_name = f"model.layers.{layer_id}.attn.k_norm.layernorm.weight"
         k_layernorm_output = f"{k_layernorm_name}/output_0"
-        self.make_external_tensor((attention.k_norm.weight.detach().cpu().to(self.to_torch_dtype[self.io_dtype]) + self.layernorm_attrs["add_offset"]).contiguous(), k_weight_name)
-        self.make_node("SimplifiedLayerNormalization", inputs=[k_reshape_1_output, k_weight_name], outputs=[k_layernorm_output], name=k_layernorm_name, **layernorm_kwargs)
-        self.make_value_info(k_layernorm_output, dtype=self.io_dtype, shape=['batch_size', 'sequence_length * num_key_value_heads', self.head_size])
+        self.make_external_tensor((attention.k_norm.weight.detach().cpu().to(new_torch_dtype) + self.layernorm_attrs["add_offset"]).contiguous(), k_weight_name)
+
+        # Create Cast nodes for inputs and outputs if old_dtype != new_dtype
+        k_layernorm_inputs = [k_reshape_1_output, k_weight_name]
```
Comment (nenad1002) [inline src/python/py/models/builder.py:1403]: q and k layernorms can easily can go in a common method.

**Review by nenad1002 — DISMISSED** [review]

**Review by nenad1002 — APPROVED** [review]
